comments for tyler (#184)
liftbridge-io/liftbridge: documentation/concepts.md

concept of a *stream*, which is a durable message stream attached to a NATS
subject. A stream consists of one or more *partitions*, which are ordered,
replicated, and durably stored on disk and serve as the unit of storage and
parallelism in Liftbridge. [@tyler Pls elaborate on where the log is stored exactly.
This is important since users will probably want attach a k8s storage to the volume->dir, eg. s3 bucket or SSDs. This will enable statefulness when using stateful sets in k8s.]

> [@Tyler The dependency on NATS also means that future changes to the NATS project will impact the Liftbridge project. This problem will be remediated by versioning of stable releases of Liftbridge pointing to a stable release of NATS?]

Author: Then it is appropriate to mention this in somewhere is the docs. Do you want to add it here in the concepts section or somewhere else. You are the captain of this boat.

off the subject and executing the command. Subsequently, the result of this activity is then posted on another subject, perhaps for down streams analytical 
reporting purposes etc. This enables the Query in the CQRS patterns. A careful reader saw the above remark of replication factor, redundancy and groups. 
These choices will impact this example with microservice workers, since the order of messages and guarantees on replication will be affected by these chioces. More on that later.

For further info on CQRS and event sourceing pls
see https://martinfowler.com/bliki/CQRS.html
and https://martinfowler.com/eaaDev/EventSourcing.html
respectively. 
```

There can be multiple streams attached to the same NATS subject, but stream
names must be unique within a cluster. 

[@Tyler Please describe if this is intended architecture or just happened due to dependecy towards NATS. If the latter we need to advice on usage. If the former, we need to descibe the use case. ]
```
This means that the concept of stream name is now defined as being the least common denominator, not the NATS subject as when using NATS alone without Liftbridge.

> ```

Author: I might have misunderstood, but how I interpret the text is as follows. Since Liftbridge can have streams aggregate 1..N NATS subjects and 1 NATS subject can contain 1..N Liftbridge streams we need advise on why one would choose a particular setup. This means one set up/configuration per use case. Also, one should point out choices of configuration that one should not use. Imagine the Liftbridge user reading the text. Identifies his/her use case and get an advise on architectural setup. That is real value. Do you agree?

By default, streams have a single partition. This partition maps directly to
the stream's NATS subject. If a stream has multiple partitions, each one maps
to a different NATS subject derived from the stream subject. For example, if a
stream with three partitions is attached to the subject "foo", the partitions
will map to the subjects "foo", "foo.1", and "foo.2", respectively. (Please note the naming convention on these subjects linked to partitions.)

Each partition has its own message log, leader, and set of followers. To reduce
resource consumption, partitions can be paused. Paused partitions are
subsequently resumed once they are published to.

```
Use case note:

> Since each partition has its own message log, it means that to meet a data lineage requirement one could point all these logs to a general log like a Linux system log by subscribing to * or concatenating the logs on the storage device attached. Aggregated or not, these logs could then be consumed by a supervisor system, e.g Logstash, Jaeger, Zipkin etc for DevOps reasons. Or just stored in an S3 bucket for persistence like a ledger of what has been communicated as events.

Author: This is not for the regular consumer. (Actually this consumer could be treated as a regular consumer but one would probably avoid this) This message log-consumer is present for the Ops team consuming enriched log-events from the liftbridge engine incl. metadata about events depending on flags like -log=INFO etc. They need all logs to be sieved into a log consumer engine for error handling, but also for crime prevention. Since I guess Liftbridge could be used on a non k8s platform this needs to be present somewhere. One analogy could be a bank doing Anti-money laundering applying AI or just Storm/Spark jobs on the regular subject events PLUS these Liftbridge log-events. Does it make sense?

partition and sends back an acknowledgement to publishers upon committing a
message to the log. A message is committed to the log once it has been
replicated to the partition's
[in-sync replica set (ISR)](#in-sync-replica-set-isr). [@Tyler: what happens when 

> replication fails? Do we lose the event?] 

Author: Please add what you just said to the text. It is vital info, since I will not be the only one asking this question. The proposal for text could be "Is it possible we can loose events because of this?  The answer is - it depends on how things are configured. If the message is published with the LEADER ack policy (default), an ack is sent back to the client as soon as the leader has stored the message. If the message is published with the ALL ack policy, the ack is sent only after all members of the ISR have stored it. Thus, if the ISR is unavailable, no ack will be sent indicating there's no guarantee the message was committed. A minimum ISR size can also be configured to provide a high level of durability, but this creates a trade-off with availability (see docs here)."

> replication fails? Do we loose the event?] 

Author: So many typos... Thanks for correcting them! 
Ipad-texting from bed it not the future for me I just realised...
:flushed:

> replication fails? Do we loose the event?] 

Reviewer: 😂 Well put. No worries, glad to be of help. Your input will greatly improve this document.

Consumers read committed messages from the log through a subscription on the
partition. They can read back from the log at any arbitrary position, or
*offset*. Additionally, consumers can wait for new messages to be appended
to the log.

```
Use case note:

> And the interested reader now identifies a typical consumer to be a stateless micro service worker. The *offset* parameter is of special interest should one

Author: The consumer must track it's offset and persist it. You are completely correct - it is a state. But is is not depended on an aggregation of events leading to a state, but the offset itself is the state needed. So please feel free to reformulate this - perhaps as "Consumers need to track their state, i.e. the offset, at least until we provide support for durable consumer groups."

"foo.*". Each of these streams will receive a copy of the message as NATS
handles this fan-out.

> [@Tyler - pls describe use case or advise on architectural implementation for the user]

Author: Does this answer relate to your questions regarding "Comment on lines +66 to +67" above? We do not need to incorporate log of the Liftbridge application itself into the audit log? Do we not risk loosing security related log entries and application glitches?

just a single partition, but multiple partitions can be created for increased
parallelism. Messages can then be delivered to partitions based on their key,
in a round-robin fashion, randomly, or with some other partitioning strategy
on the client. 

```
Use case note:
Please note that ordering within the partion is upheld but not over partitions. This means that the partioning strategy is of high importance, since an aggregated state cannot be achieved for a path dependend consumer subscribing to many streams (but perhaps on the same NATS subject). I.e. a consumer needing correct order of events can only subscribe to many streams if and only if the events are uncorrelated/independend between the partitions (and thus have order within a partition.) An architect would pay particular interest in making sure of independend and stateless workers when applying DDD. Link of interest: https://dddcommunity.org/book/evans_2003/
```
Additionally, streams can join a named load-balance group, which load balances
messages on a NATS subject amongst the streams in the group. Load-balance
groups do not affect message delivery to other streams not participating in
the group.

> [@Tyler - pls descibe use case or advise set up on architecture impementatiob for user]

Reviewer: Same as above. 😅

> [@Tyler - pls descibe use case or advise set up on architecture impementatiob for user]

Reviewer 2: Load-balance groups are for distributing messages from a subject amongst a group of streams. Note that this could also be achieved using a _partitioned_ stream with a random or round-robin partitioning strategy. However, partitioning assumes you've already "bought in" to Liftbridge streams as a first-class citizen. But imagine a case where there is a pre-existing NATS subject that Liftbridge-agnostic services are already publishing to, and we want to turn that NATS subject into a durable log that Liftbridge-aware services can consume. We could attach a single stream to the subject, but if it's very high volume, we might need a load-balance group to distribute the load across a set of streams.

usability. However, this is akin to other similar systems, like Kafka, where
you must first create a topic and then you publish to that topic.

```

> The typcal use case for a producer not caring if the ack is returned or not is a IoT device or a sensor. This means that for the sensor it is not important to know if Liftbridge indeed get to record the event. For a more regulated system, one could "assume" acknowledgements are important to the producer since the recorded truth now resides within Liftbridge, ref event sourcing.

Reviewer: ```suggestion
The typical use case for a producer not caring if the ACK is returned or not is an IoT device or a sensor. This means that for the sensor it is not important to know if Liftbridge indeed got to record the event. For a more regulated system, one could "assume" acknowledgements are important to the producer since the recorded truth now resides within Liftbridge, ref event sourcing.
```

> 
> ```
> The typical use case for a producer not caring if the ACK is returned or not is an IoT device or a sensor. This means that for the sensor it is not important to know if Liftbridge indeed got to record the event. For a more regulated system, one could "assume" acknowledgements are important to the producer since the recorded truth now resides within Liftbridge, ref event sourcing.
> ```

Reviewer 2: ```suggestion
> **Use Case Note**
>
> The typical use case for a producer not caring if the ack is returned or not is an IoT device or a sensor. This means that for the sensor, it is not important to know if Liftbridge indeed got to record the event. For a more regulated system, one could assume acknowledgements are important to the producer since the recorded truth now resides within Liftbridge, as is the case in an event-sourced system.
```

to pick up where it left off and provide fault-tolerant consumption of
streams.

> ```
> This ties back to the reporting worker starved to death, but clinging on to an *offset* to the bitter end so as not to lose probable state. When stateful consumer groups are implemented the reporting worker will not be restarted without state, but can resume from where it left off.
> ```

Reviewer 2: ```suggestion
> **Architect's Note**
>
> This ties back to the reporting worker starved to death but clinging on to an *offset* to the bitter end so as not to lose probable state described above. When stateful consumer groups are implemented, the reporting worker can be restarted without state but can resume from where it left off due to state stored by the server.
```

> This ties back to the reporting worker starved to death, but clinging on to an *offset* to the bitter end not to loose probable state. When stateful consumer groups are implemented the reporting worker need not to be restarted with no state, but only to be initiated with it's state and it can resume from there.

Reviewer: ```suggestion
This ties back to the reporting worker starved to death, but clinging on to an *offset* to the bitter end so as not to lose probable state. When stateful consumer groups are implemented the reporting worker will not be restarted without state, but can resume from where it left off.
```

`Probable` in the sentence above doesn't make sense to me. Do you mean mutable or unsaved state? 🤔

Author: probable state - this is since there is the small but not insignificant event that the consumer has time time to complete the work, but not enough time to save its own state (assuming thats the order the consumer is implemented. Perhaps there is some atomic behavior one could use - but I do not know of it)

configured to compact by key. In this case, it retains only the last message
for each unique key. Messages that do not have a key are always retained.

> ```
> From an architectural point of view, the choice here is to compact as much as possible without losing state (aggregation of events). Lineage is taken care of by the log stream if stored in e.g. an S3 bucket as noted above.
> ```

Reviewer 2: ```suggestion
> **Architect's Note**
>
> From an architectural point of view, the choice here is to compact as much as possible without losing state (aggregation of events). Lineage is taken care of by the log stream if stored, for example, in an S3 bucket as noted above.
```

Controller is also referred to as "metadata leader" in some contexts.

> [@taylor - guidance is needed on the number of deployed constrollers and possibly the on where they are deployed]

Reviewer 2: There is only a single controller (i.e. leader) at a given time which is elected by the Liftbridge cluster. Guidance on cluster size depends, but one important point here is that, currently, all servers in the cluster participate in the Raft consensus group. This has implications on the scalability of the cluster control plane, which there is an issue open to address [here](https://github.com/liftbridge-io/liftbridge/issues/41).

General advice is to run an odd number of servers in the cluster, e.g. 3 or 5, depending on scaling needs. Ideally, cluster members are run in different availability zones or racks for improved fault-tolerance.

Author: Excellent! Perhaps you could include this. I am sure user would like to see exactly this, especially coming from you.

serializing their messages into [*envelopes*](./envelope_protocol.md). An
envelope allows publishers to set things like the `AckInbox`, `Key`, `Headers`,
and other pieces of metadata.

> ```
> Do not underestimate the need for meta data on your processes. Proper enrichment of meta data will enable future business models served as new products to new
> consumers, e.g. consumer behaviour etc.
> ```

Reviewer 2: ```suggestion
> **Architect's Note**
>
> Do not underestimate the need for metadata on your processes. Proper enrichment of metadata will enable future business models served as new products to new consumers, e.g. consumer behavior, etc.
```

Author: This is not really relating to Liftbridge, but more to the business model evolving over time. The concept is as follow. You starve you competitors by offering the main product for free. You add enough meta data to the main product to enable surrounding add-on products that they are now willing to pay for, since your product is for free with great extra value added (with some additional cost which is equal to what yo gave up on the main product). The competition is gone and your consumer base is larger providing a better revenue base for you. I feel it is an important feature not only for technical meta data but also for product improvment. Do you agree? If not just delete this comment

> A final note is to read "Designing Event-Driven Systems" by Ben Stopford that can be found on confluent.io for downloading

Reviewer 2: Is there a direct link we can provide?

Reviewer: It can be requested here: https://www.confluent.io/designing-event-driven-systems/.

Still, seems a bit fishy to put a link to Confluent here which is the company behind Kafka as a Service? 😅

Author: As of now Liftbridge wants to be perceived as a better cloud native substitute for Kafka (in my eyes). Then one "pays tribute" to them. At a pivoting point people realise that Liftbridge is actually much better than Kafka, perhaps v2. Then one stops to compare oneself with Kafka. And that link is a good link. Have you read the book?

Reviewer: Not yet but I pushed it to my Kindle. 😅

Author: There is a better link. http://www.benstopford.com/2018/04/27/book-designing-event-driven-systems/

Reviewer: Great, that's much better!
